lib/ecto/query/builder/from.ex

```diff
@@ -145,6 +140,37 @@ defmodule Ecto.Query.Builder.From do
   def prefix!(prefix) when is_binary(prefix) or is_nil(prefix), do: prefix
   def prefix!(prefix), do: raise("`prefix` must be a string, got: #{inspect(prefix)}")
 
+  @doc """
+  Validates hints at compile time and runtime
+  """
+  def hint!(hint) when is_binary(hint), do: hint
+
+  def hint!({:unsafe_fragment, _, [fragment]}) do
+    case fragment do
+      {:^, _, [value]} ->
+        quote do
+          checked = Ecto.Query.Builder.From.hint!(unquote(value))
+          {:unsafe_fragment, checked}
+        end
+
+      hint when is_binary(hint) ->
+        hint
+
+      other ->
+        Builder.error!(
+          "`hints` must be a compile time string, an unsafe fragment of the form `unsafe_fragment(^...)`, " <>
+            "or a list containing either, got: `#{Macro.to_string(other)}`"
```

Comment:
```suggestion
            "or a compile time list containing either, got: `#{Macro.to_string(other)}`"
```

Reply:
I think the initial "compile time" is meant to qualify them all. :)

Reply:
Then the `an` and `a` need to be changed

Reply:
Would it read better if it was like this:

> a (list of) compile-time strings or unsafe fragments of the form `unsafe_fragment(^...)`, got: ...

Reply:
actually nm I get what you're saying now. fixed.